crates/bevy_ecs/src/query/filter.rs

```diff
@@ -512,6 +512,191 @@ macro_rules! impl_tuple_query_filter {
+/// ```
+///
+/// Filtering for components that were mutated:
+/// ```
+/// # use bevy_ecs::component::Component;
+/// # use bevy_ecs::entity::Entity;
+/// # use bevy_ecs::query::Added;
+/// # use bevy_ecs::query::Changed;
+/// # use bevy_ecs::query::Not;
+/// # use bevy_ecs::system::IntoSystem;
+/// # use bevy_ecs::system::Query;
+/// #
+/// # #[derive(Component)]
+/// # struct MyComponent;
+/// #
+/// fn print_mutated_component_system(
+///     query: Query<Entity, (Changed<MyComponent>, Not<Added<MyComponent>>)>
+/// ) {
+///     for entity in &query {
+///         println!("Entity {:?} had MyComponent changed, but was not just added", entity);
+///     }
+/// }
+/// # bevy_ecs::system::assert_is_system(print_mutated_component_system);
+/// ```
+pub struct Not<T>(PhantomData<T>);
+
+/// SAFETY:
+///
+/// For purely archetypal filters, this inverts the access in `update_component_access` as well as in
+/// `matches_component_set`, ensuring that they match.
+///
+/// For purely non-archetypal filters, this maintains the same read accesses as the original filter in
+/// both instances.
+
+unsafe impl<T> WorldQuery for Not<T>
+where
+    T: QueryFilter + InvertibleFilter,
+{
+    type Item<'a> = T::Item<'a>;
+
+    type Fetch<'a> = T::Fetch<'a>;
+
+    type State = T::State;
+
+    fn shrink<'wlong: 'wshort, 'wshort>(item: Self::Item<'wlong>) -> Self::Item<'wshort> {
+        T::shrink(item)
+    }
+
+    #[inline]
+    unsafe fn init_fetch<'w>(
+        world: UnsafeWorldCell<'w>,
+        state: &Self::State,
+        last_run: Tick,
+        this_run: Tick,
+    ) -> Self::Fetch<'w> {
+        T::init_fetch(world, state, last_run, this_run)
+    }
+
+    const IS_DENSE: bool = T::IS_DENSE;
+
+    #[inline]
+    unsafe fn set_archetype<'w>(
+        fetch: &mut Self::Fetch<'w>,
+        state: &Self::State,
+        archetype: &'w Archetype,
+        table: &'w Table,
+    ) {
+        T::set_archetype(fetch, state, archetype, table);
+    }
+
+    #[inline]
+    unsafe fn set_table<'w>(fetch: &mut Self::Fetch<'w>, state: &Self::State, table: &'w Table) {
+        T::set_table(fetch, state, table);
+    }
+
+    #[inline(always)]
+    unsafe fn fetch<'w>(
+        fetch: &mut Self::Fetch<'w>,
+        entity: Entity,
+        table_row: TableRow,
+    ) -> Self::Item<'w> {
+        T::fetch(fetch, entity, table_row)
+    }
+
+    fn update_component_access(state: &Self::State, access: &mut FilteredAccess<ComponentId>) {
+        let mut intermediate = FilteredAccess::<ComponentId>::default();
+        T::update_component_access(state, &mut intermediate);
+        intermediate.invert_filters();
+
+        access.extend(&intermediate);
+    }
+
+    fn update_archetype_component_access(
+        state: &Self::State,
+        archetype: &Archetype,
+        access: &mut Access<ArchetypeComponentId>,
+    ) {
+        T::update_archetype_component_access(state, archetype, access);
```

Comment:
I didn't find time to look for a failing example, but this feels wrong, it should be inverted (but again only for archetypal filters...)